oclc/classify.py

```diff
@@ -26,15 +30,284 @@
+            tags = cls._xpath(tree, "//oclc:%s" % name)
+            if tags:
+                mapping[name] = tags
+        return mapping
+
+    @classmethod
+    def get_subjects(cls, classifiers):
+        """Convert a list of tags from an XML document into a corresponding
+        list of SubjectData objects.
+        
+        :param classifiers: A dictionary mapping classifier names to
+        etree tags. The classifier name is the type of classifier--DDC, LCC, or
+        Fast.
+
+        The element tags are formatted such that the names are buried
+        in a longer string (e.g. "{http://classify.oclc.org}ddc"), so
+        it's easiest to just keep track of them separately, rather
+        than having to extract them every time.
+        """
+        subjects = []
+        for classifier_name, classifier_tags in classifiers.items():
+            for tag in classifier_tags:
+                subtags = cls._extract_subtags(classifier_name, tag)
+                subjects.extend(
+                    cls.make_subject_data(classifier_name, subtags)
+                )
+        return subjects
+
+    @classmethod
+    def _extract_subtags(cls, classifier_name, classifier_tag):
+        """
+        :param classifier_name: the name of a classifier type (e.g. "ddc")
+        :param classifier_tag: An etree tag containing classification
+           data.
+        """
+
+        # Some classifier types use 'heading', some use 'mostPopular'.
+        # Try it both ways and see which one works.
+        matches = None
+        for item in ("heading", "mostPopular"):
+            search_term = "//oclc:%s//oclc:%s" % (classifier_name, item)
+            matches = cls._xpath(classifier_tag, search_term)
+            if matches:
+                break
+        return matches
+
+    @classmethod
+    def make_subject_data(cls, classifier_name, classifier_tags):
+        """Create a SubjectData object for every tag.
+
+        :param classifier_name: The name of the classifier to use
+        when determining what this SubjectData means.
+
+        :param classifier_tags: A list of etree tags, each containing
+        a single classification.
+        """
+        results = []
+        for tag in classifier_tags:
+            weight, identifier, name = cls._parse_subject_tag(tag)
+            subject_data = SubjectData(
+                type=cls.CLASSIFIERS[classifier_name],
+                weight=weight,
+                identifier=identifier,
+                name=name,
+            )
+            results.append(subject_data)
+        return results
+
+
+    @classmethod
+    def _parse_subject_tag(cls, tag):
+        """Extract from an etree tag the information necessary to
+        create a SubjectData.
+        """
+        # The names of the attributes vary depending on which type of classifier
+        # the tag is for.  Only Fast tags have a name.
+        weight = tag.get("holdings") or tag.get("heldby")
```

Comment:
I'm not sure how the weight is used but it seems like it would need to be normalized to be useful. Would it make sense to divide by the total holdings for the work?

Reply:
I hadn't thought about this, but having thought about it, I think leaving it as-is is the best solution.

The weights measure the strength of the evidence. We grab all the Classifications for an Identifier (and other Identifiers equivalent to that one), put them all in a weighted data structure, and see which genres are the most strongly implied. So if 1000 libraries classify a book as fiction and 10 as nonfiction, then fiction wins.

If there's only one Identifier (which is going to be the case most of the time, until we bring OCLC Linked Data back online) then the absolute numbers don't matter. 1000 vs. 10 is the same as 1 versus 0.001.

But then let's say we bring in data from another source like OCLC Linked Data. The Linked Data subjects aren't weighted, so they all end up with a weight of 1. So for most ISBNs, the classifications from OCLC Classify are going to swamp the classifications from OCLC Linked Data.

In one sense that's not fair to OCLC Linked Data. But knowing that 1000 libraries classified a book a certain way is really strong evidence. OCLC Linked Data doesn't provide that evidence, so it can only help on the margins, or when OCLC Classify data is unavailable.

This is a big part of the reason I disabled the OCLC Linked Data coverage provider -- it was doing a lot of work little benefit. I'm hoping that OCLC Classify + VIAF will give us almost all the benefits of the old system at a much lower cost.